e2e-tests/test/main.test.ts

```diff
@@ -1,69 +1,121 @@
+    const action = async () => {
+      const wallet = new Wallet(RichAccounts[0].PrivateKey, provider);
+      const deployer = new Deployer(hre, wallet);
+
+      // setup user wallet with 3 ETH
+      const userWallet = Wallet.createRandom().connect(provider);
+      await wallet.sendTransaction({
+        to: userWallet.address,
+        value: ethers.utils.parseEther("3"),
+      });
+
+      // deploy Greeter contract
+      const artifact = await deployer.loadArtifact("Greeter");
+      const greeter = await deployer.deploy(artifact, ["Hello, world!"]);
+
+      // should revert
+      const tx = await greeter.connect(userWallet).setGreeting("Hola, mundo!");
+      await tx.wait();
+    };
+
+    await expectThrowsAsync(action, "Ownable: caller is not the owner");
+  });
+
+  it("Should produce event logs", async function () {
+    const wallet = new Wallet(RichAccounts[0].PrivateKey);
+    const deployer = new Deployer(hre, wallet);
+
+    const greeter = await deployContract(deployer, "Greeter", ["Hi"]);
+
+    expect(await greeter.greet()).to.eq("Hi");
+
+    const setGreetingTx = await greeter.setGreeting("Luke Skywalker");
+    const receipt: TransactionReceipt = await setGreetingTx.wait();
+
+    // Validate log is created
+    expect(receipt.logs.length).to.greaterThanOrEqual(1);
+    const setGreetingLog = receipt.logs[1];
+    expect(setGreetingLog.address).to.equal(greeter.address);
+    const normalizedSetGreetingLogData = ethers.utils
+      .toUtf8String(setGreetingLog.data)
+      .replaceAll("\u0000", "")
+      .replace(" +", "");
+    expect(normalizedSetGreetingLogData).to.equal(
+      "Greeting is being updated to Luke Skywalker"
+    );
+  });
+
+  it("Should filter event logs", async function () {
+    const wallet = new Wallet(RichAccounts[0].PrivateKey);
+    const deployer = new Deployer(hre, wallet);
+    const greeter = await deployContract(deployer, "Greeter", ["Hi"]);
+    expect(await greeter.greet()).to.eq("Hi");
+
+    const setGreetingTx = await greeter.setGreeting("Luke Skywalker");
+    let receipt: TransactionReceipt = await setGreetingTx.wait();
+
+    // Create filter
+    const topic = receipt.logs[1].topics[0];
+    const filterId = await provider.send("eth_newFilter", [
+      {
+        fromBlock: "earliest",
+        toBlock: "latest",
+        topics: [topic],
+      },
+    ]);
+
+    // New filter should be empty
+    let filterChanges: Log[] = await provider.send("eth_getFilterChanges", [
+      filterId,
+    ]);
+    expect(filterChanges).to.empty;
+
+    // Emit logs and filter should not be empty
+    receipt = await (await greeter.setGreeting("Darth Vader")).wait();
+    filterChanges = await provider.send("eth_getFilterChanges", [filterId]);
+
+    expect(filterChanges.length).to.eq(1);
+    expect(filterChanges[0].transactionHash).to.eq(receipt.transactionHash);
+    expect(filterChanges[0].blockHash).to.eq(receipt.blockHash);
+    const normalizedSetGreetingLogData = ethers.utils
```

Comment:
can you add a comment about this? why do we need this?

Reply:
Updated to use interfaces to avoid string normalization